clippy_lints/src/undocumented_unsafe_blocks.rs

```diff
@@ -542,32 +627,70 @@ fn stmt_has_safety_comment(cx: &LateContext<'_>, span: Span, hir_id: HirId) -> H
     HasSafetyComment::Maybe
 }
 
-fn comment_start_before_item_in_mod(
+trait CheckableItemId {
+    fn owner_id(&self) -> hir::OwnerId;
+}
+
+impl CheckableItemId for hir::ItemId {
+    fn owner_id(&self) -> hir::OwnerId {
+        self.owner_id
+    }
+}
+impl CheckableItemId for hir::ImplItemRef {
+    fn owner_id(&self) -> hir::OwnerId {
+        self.id.owner_id
+    }
+}
+impl CheckableItemId for hir::TraitItemRef {
+    fn owner_id(&self) -> hir::OwnerId {
+        self.id.owner_id
+    }
+}
```

Comment:
An option would be to have `ids` be an iterator rather than a slice, and find the previous node with `ids.take_while(|id| id != owner_id).last()`, if it's `None` then you're in the first entry

May or may not be nicer than the traits + indexing into the slice, so feel free to keep it as is if you prefer it as is

Reply:
Seems better and a lot shorter, thanks, although the logic is harder to understand, took my dumb brain a while to figure out what is happening and why it works :rofl:

Reply:
The current `enumerate().find_map()` finds the index of the current id then subtracts 1 to get the index of the previous node, may be clearer as ~

```rust
let idx = ids.iter().pos(|id| id.owner_id() == owner.def_id())?;
if idx == 0 {
    walk_span_to_context(search_span, SyntaxContext::root()).map(Span::lo)
} else {
    let span = cx.tcx.hir().span(ids[idx - 1]);
    walk_span_to_context(span, SyntaxContext::root()).map(Span::lo)
}
```

The `.take_while().last()` finds the id *preceding* the current one so doesn't require that re-indexing. If it's `None` then the current id was the first in `siblings`, since the `take_while` would return an iterator of 0 items


To make it clearer you could have it be `if let Some(prev_sibling_id) = siblings...`

Reply:
> The `.take_while().last()` finds the id _preceding_ the current one so doesn't require that re-indexing...

Thanks, wouldn't the preceding id be a one that `!= owner_id` tho? I guess what gets me confused is that it was previously accepting only when it `== owner_id`

Reply:
Say you have an iterator of `['a', 'b', 'c', 'd', 'e']`, `.take_while(|n| n != 'd')` would give you

```text
['a', 'b', 'c', 'd', 'e']
 ^^^^^^^^^^^^^ an iterator of these three items
```

The `.last()` would then give you the `3`

Feel free to use a different way to find the previous item if you think of a clearer one though, another way could be to make it a `DoubleEndedIterator` and do

```rust
siblings.rfind(|id| id == owner_id);
if let Some(prev_sibling_id) = siblings.next_back() {
...
```

Reply:
ahhh, that make so much more sense! thanks again for your patience. 

Also I ran a simple bench with these two methods and found out that using `DoubleEndedIterator` is faster, so I think I'm gonna use that